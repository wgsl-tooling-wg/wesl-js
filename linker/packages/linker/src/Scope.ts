import { DeclarationElem } from "./AbstractElems2.ts";

export interface SrcModule {
  /** module path "rand_pkg::sub::foo", or "package::main" */
  modulePath: string;
  
  /** file path to the module for user error reporting e.g "rand_pkg:sub/foo.wesl", or "./sub/foo.wesl" */
  filePath: string;

  /** original src for module */
  src: string;
}

export interface SrcLoc {
  pos: number;
  src: SrcModule;
}

/** a src declaration or reference to an ident */
export type Ident = DeclIdent | RefIdent;

export type Conditions = Record<string, boolean>;

interface IdentBase {
  originalName: string; // name in the source code for ident matching (may be mangled in the output)
  conditions?: Conditions; // conditions under which this ident is valid (combined from all containing elems)
}

export interface RefIdent extends IdentBase {
  kind: "ref";
  refersTo?: Ident; // import or decl ident in scope to which this ident refers. undefined before binding
  std?: true; // true if this is a standard wgsl identifier (like sin, or u32)
}

export interface DeclIdent extends IdentBase {
  kind: "decl";
  mangledName?: string; // name in the output code
  declElem: DeclarationElem;
}

export type ScopeKind =
  | "module" // root scope for a module (file)
  | "body"; // a scope inside the module (fn body, nested block, etc.)

/** tree of ident references, organized by lexical scope */
export interface Scope {
  id?: number; // for debugging
  idents: Ident[]; // idents found in lexical order in this scope
  parent: Scope | null; // null for root scope in a module
  children: Scope[];
  kind: ScopeKind;
}

export function resetScopeIds() { // for debugging
  scopeId = 0;
}

let scopeId = 0; // for debugging

/** make a new Scope object */
export function makeScope(s: Omit<Scope, "id">): Scope {
  return { ...s, id: scopeId++ };
}

export interface RootAndScope {
  rootScope: Scope;
  scope: Scope;
}

export function emptyScope(kind: ScopeKind): Scope {
  return makeScope({ idents: [], parent: null, children: [], kind });
}

export function emptyBodyScope(parent: Scope): Scope {
  return makeScope({ kind: "body", idents: [], parent, children: [] });
}

/** For debugging,
 * @return true if a scope is in the rootScope tree somewhere */
export function containsScope(rootScope: Scope, scope: Scope): boolean {
  if (scope === rootScope) {
    return true;
  }
  for (const child of rootScope.children) {
    if (containsScope(child, scope)) {
      return true;
    }
  }
  return false;
}

export function exportDecl(scope:Scope, name: string): DeclIdent | undefined {
  for (const ident of scope.idents) {
    if (ident.originalName === name && ident.kind === "decl") {
      return ident;      
    }
  }
}