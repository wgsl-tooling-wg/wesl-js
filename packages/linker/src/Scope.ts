import { pretty } from "berry-pretty";
import { tracing } from "mini-parse";

type IdentKind = "decl" | "ref";

export interface SrcModule {
  /** full path to the module e.g "package/sub/foo", or "_root/sub/foo" */
  modulePath: string;

  /** original src for module */
  src: string;
}

export interface SrcLoc {
  pos: number;
  src: SrcModule;
}

/** a src declaration or  */
export interface Ident {
  kind: IdentKind;
  refersTo?: Ident | null; // preceding ident in scope. null for decls, undefined before binding
  originalName: string; // name in the source code for ident matching (may be mangled in the output)
}

export type ScopeKind =
  | "module" // root scope for a module (file)
  | "body"; // a scope inside the module (fn body, nested block, etc.)

/** tree of ident references, organized by lexical scope */
export interface Scope {
  idents: Ident[]; // idents found in lexical order in this scope
  parent: Scope | null; // null for root scope in a module
  children: Scope[];
  kind: ScopeKind;
}

export interface RootAndScope {
  rootScope: Scope;
  scope: Scope;
}

/** @return a new root scope and new current scope with an ident added to the current scope. */
export function withAddedIdent(
  rootScope: Scope,
  scope: Scope,
  ident: Ident,
): RootAndScope {
  if (tracing && !containsScope(rootScope, scope)) {
    logScope("scope not in rootScope", rootScope);
    logScope("..scope", rootScope);
  }

  // clone the current provisional scope with new ident added
  const scopeIdents = scope.idents;
  const idents: Ident[] = [...scopeIdents, ident];
  const newScope: Scope = { ...scope, idents };
  const newRootScope = cloneScopeReplace(rootScope, scope, newScope);

  return {
    scope: newScope,
    rootScope: newRootScope,
  };
}

/** @return a new root scope and new current scope with a child scope added to the current scope. */
export function withChildScope(
  rootScope: Scope,
  currentScope: Scope,
  newChildScope: Scope,
): RootAndScope {
  const newCurrentScope = {
    ...currentScope,
    children: [...currentScope.children, newChildScope],
  };
  const newRootScope = cloneScopeReplace(
    rootScope,
    currentScope,
    newCurrentScope,
  );

  return {
    scope: newChildScope,
    rootScope: newRootScope,
  };
}

/** For debugging,
 * @return true if a scope is in the rootScope tree somewhere */
function containsScope(rootScope: Scope, scope: Scope): boolean {
  if (scope === rootScope) {
    return true;
  }
  for (const child of rootScope.children) {
    if (containsScope(child, scope)) {
      return true;
    }
  }
  return false;
}

/** clone the rootScope, replacing oldScope with newScope in child and parent links */
function cloneScopeReplace(
  rootScope: Scope,
  oldScope: Scope,
  newScope: Scope,
): Scope {
  if (rootScope === oldScope) {
    return newScope;
  }

  const { kind, idents } = rootScope;
  const { parent: rootParent, children: rootChildren } = rootScope;
  const parent = rootParent === oldScope ? newScope : rootParent;
  const children = rootChildren.map(s =>
    cloneScopeReplace(s, oldScope, newScope),
  );
  return { kind, idents, parent, children };
}

export function logScope(message: string, scope: Scope) {
  console.log(`${message}:`);
  console.log(scopeToString(scope, 2));
}

export function scopeToString(scope: Scope, indent = 0): string {
  const { kind, idents, children } = scope;
  const identStr = pretty(idents.map(i => i.originalName));
  const childStrings = children.map(c => scopeToString(c, indent + 4));
  const childrenStr = childStrings.join("\n");
  const spc = " ".repeat(indent);
  // prettier-ignore
  return `${spc}${kind} ${identStr}\n` + 
         `${spc}  children:\n`+ 
         `${childrenStr}`;
}

// export interface ProvisionalScope {
//   currentScope: Scope; // current scope at this level of parsing
//   idents: Ident[]; // idents pending to add to current scope
//   scopes: Scope[]; // scopes pending to add to current scope

// }

// TODO do we need to bother with the tricky indexing to symbols?
// . two level table seems relevant for multithreading..
//   . also useful for say watch mode
//   . lets do it
// . and why do we need to use indices into the table vs pointers?
//   . if we want to replace a symbol we could just mutate it in place
//   . (esbuild claims its handy to clone the symbol table, so index references stay valid)
//   . lets skip it til we know we need it
// . and why do we need a single global symbol table to collect all the symbols
//   . scope tables seem sufficient..
//   . lets skip it til we know we need it
//
// let's build some stuff and find out..

/* 
Key tasks:
- accumulate idents and scopes during the parse
- bind idents references to declarations
- traverse AST from root module to emit
  - store idents in the AST? and references to AST from idents?
  - That would let us traverse the AST to find referenced AST elements
    - e.g. fn foo() { bar();} we need to traverse to bar()

Datastructures should be compatible for future parallelization:
- parse files in parallel
  - each file gets its own ident table
- presume ident binding is single threaded for now
  - esbuild is single threaded (according to arch diagram)
  - do any necessary ident mangling
  - works entirely on ident table
- emit code in parallel
  - 

*/

// export interface AllIdents {
//   tables: IdentTable[];   // idents for all tables
// }

// export interface IdentTable {
//   idents: Ident[][]; // indexed by moduleIndex and identIndex
// }

// // reference to a symbol in the Ident table, e.g. from an AST
// export interface IdentRef {
//   moduleIndex: number; // index of the file
//   identIndex: number; // index of the identifier within the file
// }
