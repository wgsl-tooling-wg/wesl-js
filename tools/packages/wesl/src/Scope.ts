import { FunctionParam, GlobalDeclarationElem } from "./AbstractElems.ts";

export interface SrcModule {
  /** module path "rand_pkg::sub::foo", or "package::main" */
  modulePath: string;

  /** file path to the module for user error reporting e.g "rand_pkg:sub/foo.wesl", or "./sub/foo.wesl" */
  debugFilePath: string;

  /** original src for module */
  src: string;
}

export type Conditions = Record<string, boolean>;

export interface RefIdent {
  kind: "ref";
  originalName: string;
  /** import or decl ident in scope to which this ident refers. undefined before binding */
  refersTo?: DeclIdent;
  /** true if this is a standard wgsl identifier (like sin, or u32) */
  std?: true;
}

export interface DeclIdent {
  kind: "decl";
  originalName: string;
  /** name in the output code */
  mangledName?: string;
  /** link to AST so that we can traverse scopes and know what elems to emit // TODO make separate GlobalDecl kind with this required */
  declElem: GlobalDeclarationElem | FunctionParam;
  /** To figure out which module this declaration is from. */
  srcModule: SrcModule;
}

/** tree of ident references, organized by lexical scope. */
export interface Scope {
  id?: number; // for debugging
  /** idents found in lexical order in this scope */
  idents: (RefIdent | DeclIdent)[];
  /** null for root scope in a module */
  parent: Scope | null;
  children: Scope[];
}

export function resetScopeIds() {
  // for debugging
  scopeId = 0;
}

let scopeId = 0; // for debugging

/** make a new Scope object */
export function makeScope(s: Omit<Scope, "id">): Scope {
  return { ...s, id: scopeId++ };
}

export function emptyScope(parent: Scope | null): Scope {
  return makeScope({
    idents: [],
    parent,
    children: [],
  });
}

/** For debugging,
 * @return true if a scope is in the rootScope tree somewhere */
export function containsScope(rootScope: Scope, scope: Scope): boolean {
  if (scope === rootScope) {
    return true;
  }
  for (const child of rootScope.children) {
    if (containsScope(child, scope)) {
      return true;
    }
  }
  return false;
}
