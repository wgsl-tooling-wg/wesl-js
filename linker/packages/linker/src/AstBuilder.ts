/** */

import { dlog } from "berry-pretty";
import { AbstractElem } from "./AbstractElems.ts";
import { ParsedModule } from "./Linker2.ts";
import { emptyBodyScope, Ident, makeScope, Scope, SrcModule } from "./Scope.ts";

type BuilderElem = StartElem | EndElem | StartScope | EndScope | AddIdent;

interface BuilderElemBase {
  position: number;
}

interface StartElem extends BuilderElemBase {
  kind: "elem";
  elemKind: Pick<AbstractElem, "kind">;
  start: number;
  end: number;
}

interface EndElem extends BuilderElemBase {
  kind: "endElem";
}

interface StartScope extends BuilderElemBase {
  kind: "scope";
}

interface EndScope extends BuilderElemBase {
  kind: "endScope";
}

interface AddIdent extends BuilderElemBase {
  kind: "ident";
  ident: Ident;
}

export interface HasPosition {
  /** get the current position */
  position(): number;
}

export class AstBuilder {
  #elems: BuilderElem[] = [];

  constructor(
    readonly srcModule: SrcModule,
    readonly location: HasPosition,
  ) {}

  startElem(elemKind: Pick<AbstractElem, "kind">, start: number, end: number) {
    this.addElem("elem", { elemKind, start, end });
  }

  endElem() {
    this.addElem("endElem", {});
  }

  startScope() {
    this.addElem("scope", {});
  }

  endScope() {
    this.addElem("endScope", {});
  }

  addIdent(kind: Ident["kind"], start: number, end: number) {
    const originalName = this.srcModule.src.slice(start, end);
    const ident: Ident = { kind, originalName };
    this.addElem("ident", { ident });
    // TODO create IdentElem too
  }
  
  backtrack(pos: number) {
    let i = this.#elems.length - 1;
    while (i >= 0 && this.#elems[i].position > pos) {
      i--;
    }
    this.#elems.length = i + 1;
  }

  build(): ParsedModule {
    return buildAstAndScope(this.#elems);
  }


  private addElem<T extends BuilderElem>(
    kind: T["kind"],
    fields: Omit<T, "kind"|"position">,
  ) {
    const position = this.location.position();
    const elem = { kind, position, ...fields } as T;
    this.#elems.push(elem);
    // TODO create IdentElem too
  }
}

export function buildAstAndScope(
  elems: BuilderElem[],
  scope = makeScope({ idents: [], parent: null, children: [], kind: "module" }),
  ast: AbstractElem[] = [],
): ParsedModule {
  let currentScope = scope;
  elems.forEach((elem, i) => {
    switch (elem.kind) {
      case "scope":
        const newScope = emptyBodyScope(currentScope);
        currentScope.children.push(newScope);
        currentScope = newScope;
        break;
      case "endScope":
        if (!currentScope.parent) {
          console.log(`endScope: unbalanced scope: build elem #${i}`);
          throw new Error(`endScope: unbalanced scope`);
        }
        currentScope = currentScope.parent;
        break;
      case "ident":
        currentScope.idents.push(elem.ident);
        break;
      default:
        dlog(`NYI elem kind: ${elem.kind} #${i}`);
        break;
    }
  });
  return { rootScope: scope, rootElems: ast };
}
