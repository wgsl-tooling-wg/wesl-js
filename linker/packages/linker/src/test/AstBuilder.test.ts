import { expect, test } from "vitest";
import { AstBuilder, buildAstAndScope } from "../AstBuilder.ts";
import { SrcModule } from "../Scope.ts";
import { dlog } from "berry-pretty";

test("build scope with ident", () => {
  const srcModule: SrcModule = {
    modulePath: "test.wesl",
    src: "let x = 1;",
  };
  const ast = new AstBuilder(srcModule);
  ast.addIdent("decl", 4, 5);
  const { rootScope } = ast.build();
  expect({ rootScope }).toMatchSnapshot();
});

test("build scope with child scope", () => {
  const srcModule: SrcModule = {
    modulePath: "test.wesl",
    //    012345678901234567890123456789012345678901234567890123456789012345678901234567890
    src: "fn foo() { x++; }",
  };
  const ast = new AstBuilder(srcModule);
  ast.addIdent("decl", 3, 6);
  ast.startScope();
  ast.addIdent("ref", 11, 12);
  ast.endScope();

  const { rootScope } = ast.build();
  expect({ rootScope }).toMatchSnapshot();
});
