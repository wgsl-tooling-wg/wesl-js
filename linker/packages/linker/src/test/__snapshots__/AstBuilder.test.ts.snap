// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`build scope with child scope 1`] = `
{
  "rootScope": {
    "children": [
      {
        "children": [],
        "id": 2,
        "idents": [
          {
            "kind": "ref",
            "originalName": "x",
          },
        ],
        "kind": "body",
        "parent": [Circular],
      },
    ],
    "id": 1,
    "idents": [
      {
        "kind": "decl",
        "originalName": "foo",
      },
    ],
    "kind": "module",
    "parent": null,
  },
}
`;

exports[`build scope with ident 1`] = `
{
  "rootScope": {
    "children": [],
    "id": 0,
    "idents": [
      {
        "kind": "decl",
        "originalName": "x",
      },
    ],
    "kind": "module",
    "parent": null,
  },
}
`;
